File: chain-of-responsibility/java/CorClient.java
enum RequestType {
  ConcreteHandler1, ConcreteHandler2
}

class Request {
  RequestType requestType;
  String requestDescription;

  public Request(RequestType requestType, String requestDescription) {
    this.requestType = requestType;
    this.requestDescription = requestDescription;
  }

  @Override
  public String toString() {
    return "Request{" +
        "requestType=" + requestType +
        ", requestDescription='" + requestDescription + '\'' +
        '}';
  }

  public RequestType getRequestType() {
    return requestType;
  }
}

class Handler {
  private Handler next;

  public Handler(Handler next) {
    this.next = next;
  }

  void handleRequest(Request request) {
    if (next != null) {
      next.handleRequest(request);
    }
  }
}

class ConcreteHandler1 extends Handler {

  public ConcreteHandler1(Handler next) {
    super(next);
  }

  @Override
  void handleRequest(Request request) {
    if (request.getRequestType().equals(RequestType.ConcreteHandler1))
      System.out.println("ConcreteHandler1 instance has been created");
    else
      super.handleRequest(request);
  }
}

class ConcreteHandler2 extends Handler {

  public ConcreteHandler2(Handler next) {
    super(next);
  }

  @Override
  void handleRequest(Request request) {
    if (request.getRequestType().equals(RequestType.ConcreteHandler2))
      System.out.println("ConcreteHandler2 instance has been created");
    else
      super.handleRequest(request);
  }
}

public class CorClient {
  public static void main(String[] args) {
    Handler handler = new Handler(new ConcreteHandler1(new ConcreteHandler2(null)));
    handler.handleRequest(new Request(RequestType.ConcreteHandler1, "ConcreteHandler1"));
    handler.handleRequest(new Request(RequestType.ConcreteHandler2, "ConcreteHandler2"));
  }
}

File: chain-of-responsibility/python/CorClient.py
#!/usr/bin/env python
# coding:utf8
# @Date    : 2018/9/12
# @Author  : Koon
# @Link    : chenzeping.com
# When I wrote this, only God and I understood what I was doing. Now, God only knows.

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function


class Request(object):
    def __init__(self, _type):
        self._type = _type


class Handler(object):
    def __init__(self, _next):
        self._next = _next

    def handle_request(self, request):
        self._next.handle_request(request)


class ConcreteHandler1(Handler):
    def handle_request(self, request):
        if request._type == "ConcreteHandler1":
            print("ConcreteHandler1 instance has been created")
        else:
            self._next.handle_request(request)


class ConcreteHandler2(Handler):
    def handle_request(self, request):
        if request._type == "ConcreteHandler2":
            print("ConcreteHandler2 instance has been created")
        else:
            self._next.handle_request(request)


class DefaultHandler(Handler):
    def __init__(self):
        pass

    def handle_request(self, request):
        print("Without this request type")


if __name__ == "__main__":
    handler = ConcreteHandler1(ConcreteHandler2(DefaultHandler()))
    handler.handle_request(Request("ConcreteHandler1"))
    handler.handle_request(Request("ConcreteHandler2"))
    handler.handle_request(Request("ConcreteHandler3"))
